opt: disallow mutations under union

RaduBerinde opened the pull request:
#### opt: fix With and WithScan properties

The `With` and `WithScan` property builders start by copying the input
props and then correcting things. This is error-prone: we have in time
identified multiple things that were being incorrectly inherited.

This change rewrites the code to start with empty properties and then
fill things in explicitly (like most other operators). Known problems
that are fixed by this change:
 - `CanMutate` was incorrectly set for `WithScan`
 - Rule props like `PruneCols` and `WithUses` were incorrectly
   inherited by `With`.

Release note: None

#### opt: disallow mutations under union

Because the inputs of a UNION or UNION ALL are run in parallel, it is
not possible for either of them to be mutations. The right way to do
this is to use WITH (above the union). The optimizer will have code to
pull up mutations into top-level WITHs, but until then we want to
disallow such queries from executing. This change adds a check for
this condition in the execbuilder.

Informs #40853.

Release note (sql change): Mutations under UNION or UNION ALL are
disallowed; WITH should be used on top of the union operation instead.

Release justification: low-risk fix for high-severity bug (category
3).

RaduBerinde committed:
opt: disallow mutations under union

Because the inputs of a UNION or UNION ALL are run in parallel, it is
not possible for either of them to be mutations. The right way to do
this is to use WITH (above the union). The optimizer will have code to
pull up mutations into top-level WITHs, but until then we want to
disallow such queries from executing. This change adds a check for
this condition in the execbuilder.

Informs #40853.

Release note (sql change): Mutations under UNION or UNION ALL are
now disallowed; WITH should be used on top of the union operation
instead. This restriction is temporary and will be lifted in a future
release.

Release justification: low-risk fix for high-severity bug (category
3).

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 3 of 3 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/40975#-:-LpUcHcfDxCquV4Fk2pz:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andreimatei, @jordanlewis, and @RaduBerinde)

___
*[pkg/sql/logictest/testdata/logic_test/union, line 312 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/40975#-LpUbuh18jXLqgpTaBgh:-LpUbuh18jXLqgpTaBgi:b-431qhm) ([raw file](https://github.com/cockroachdb/cockroach/blob/9773d582976fa466131fa1beab4c91158fd45ff4/pkg/sql/logictest/testdata/logic_test/union#L312)):*
> ```
> statement ok
> SELECT * FROM uniontest
> EXCEPT SELECT * FROM [INSERT INTO uniontest VALUES (1), (1) RETURNING k, v]
> ```

Maybe also add a test using a mutation inside WITH + UNION to test the non-error path?


<!-- Sent from Reviewable.io -->

justinj reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/40975)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andreimatei, @jordanlewis, and @RaduBerinde)

___
*[pkg/sql/opt/exec/execbuilder/relational.go, line 1111 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/40975#-LpUfn4Q3oAT6hyloIs0:-LpUfn4R6kFF1bZOURDf:b-7tmbme) ([raw file](https://github.com/cockroachdb/cockroach/blob/9773d582976fa466131fa1beab4c91158fd45ff4/pkg/sql/opt/exec/execbuilder/relational.go#L1111)):*
> ```Go
> 	// (because mutations can't run in parallel). Once we pull up all mutations
> 	// into top-level With expressions, this case will not be possible anymore.
> 	if typ == tree.UnionOp && (leftExpr.Relational().CanMutate || rightExpr.Relational().CanMutate) {
> ```

This could still happen today via inlining I think, though I'm not sure the best way to fix that. Maybe simplest would be to just have set ops report multiple uses of any CTEs underneath them to block inlining?


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/40975)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andreimatei, @jordanlewis, @justinj, @RaduBerinde, and @rytaft)

___
*[pkg/sql/logictest/testdata/logic_test/union, line 312 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/40975#-LpUbuh18jXLqgpTaBgh:-LpUeyYK0JqPPS_8kmAD:bouqoxr) ([raw file](https://github.com/cockroachdb/cockroach/blob/9773d582976fa466131fa1beab4c91158fd45ff4/pkg/sql/logictest/testdata/logic_test/union#L312)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Maybe also add a test using a mutation inside WITH + UNION to test the non-error path?
</blockquote></details>

Wow, such a great idea - it didn't work! Had to fix `CanMutate` for `WithScan` (CC @justinj)


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
___
*[pkg/sql/opt/exec/execbuilder/relational.go, line 1111 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/40975#-LpUfn4Q3oAT6hyloIs0:-LpUh8tnEBrbjGNUyYEo:b-8uoilo) ([raw file](https://github.com/cockroachdb/cockroach/blob/9773d582976fa466131fa1beab4c91158fd45ff4/pkg/sql/opt/exec/execbuilder/relational.go#L1111)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

This could still happen today via inlining I think, though I'm not sure the best way to fix that. Maybe simplest would be to just have set ops report multiple uses of any CTEs underneath them to block inlining?
</blockquote></details>

Are you talking about the `InlineWith` rule? That doesn't fire if `CanHazSideEffects`


<!-- Sent from Reviewable.io -->

justinj reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/40975)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andreimatei, @jordanlewis, @justinj, @RaduBerinde, and @rytaft)

___
*[pkg/sql/opt/exec/execbuilder/relational.go, line 1111 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/40975#-LpUfn4Q3oAT6hyloIs0:-LpUiUWKExMTqNPOSzRm:b-7qxevu) ([raw file](https://github.com/cockroachdb/cockroach/blob/9773d582976fa466131fa1beab4c91158fd45ff4/pkg/sql/opt/exec/execbuilder/relational.go#L1111)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Are you talking about the `InlineWith` rule? That doesn't fire if `CanHazSideEffects`
</blockquote></details>

Oh yeah, of course!


<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
> Reviewed 4 of 4 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/40975)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andreimatei, @jordanlewis, and @RaduBerinde)

___
*[pkg/sql/logictest/testdata/logic_test/union, line 312 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/40975#-LpUbuh18jXLqgpTaBgh:-LpVq2M46t6QVK1P3Evc:b-83sjoh) ([raw file](https://github.com/cockroachdb/cockroach/blob/9773d582976fa466131fa1beab4c91158fd45ff4/pkg/sql/logictest/testdata/logic_test/union#L312)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Wow, such a great idea - it didn't work! Had to fix `CanMutate` for `WithScan` (CC @justinj)
</blockquote></details>

Nice!
___
*[pkg/sql/opt/memo/logical_props_builder.go, line 754 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/40975#-LpVp9TL4SJQUG_Te4aB:-LpVp9TM2WfhooUhkdAC:bsts59q) ([raw file](https://github.com/cockroachdb/cockroach/blob/8bd8d6c134d29fbabc8e945f0d439ebe9786a77b/pkg/sql/opt/memo/logical_props_builder.go#L754)):*
> ```Go
> 	// WithScan inherits most of the logical properties of the expression it
> 	// references.
> 	*rel = *ref.BindingProps
> ```

It feels like this inheritance of properties has been a source of a lot of bugs recently. It seems like one by one we're just resetting the props below as we find and fix bugs. Maybe we should rethink this? cc @justinj


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/40975)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andreimatei, @jordanlewis, @justinj, and @RaduBerinde)

___
*[pkg/sql/opt/memo/logical_props_builder.go, line 754 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/40975#-LpVp9TL4SJQUG_Te4aB:-LpY7HY84wbKQxo7VpZC:b-e1whz8) ([raw file](https://github.com/cockroachdb/cockroach/blob/8bd8d6c134d29fbabc8e945f0d439ebe9786a77b/pkg/sql/opt/memo/logical_props_builder.go#L754)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

It feels like this inheritance of properties has been a source of a lot of bugs recently. It seems like one by one we're just resetting the props below as we find and fix bugs. Maybe we should rethink this? cc @justinj
</blockquote></details>

Yeah, we should be doing a whitelist instead of a blacklist (i.e. copy those that are ok to copy). I'll try to rework it (I'll make a separate commit in this PR).


<!-- Sent from Reviewable.io -->

justinj reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/40975#-:-LpY94AZ71O6eTkZu4Y1:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @andreimatei, @jordanlewis, @RaduBerinde, and @rytaft)

___
*[pkg/sql/opt/memo/logical_props_builder.go, line 754 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/40975#-LpVp9TL4SJQUG_Te4aB:-LpY8zph99KnLBxMbEKp:bp76cem) ([raw file](https://github.com/cockroachdb/cockroach/blob/8bd8d6c134d29fbabc8e945f0d439ebe9786a77b/pkg/sql/opt/memo/logical_props_builder.go#L754)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Yeah, we should be doing a whitelist instead of a blacklist (i.e. copy those that are ok to copy). I'll try to rework it (I'll make a separate commit in this PR).
</blockquote></details>

Yeah, agreed, this has proven to be the wrong approach.


<!-- Sent from Reviewable.io -->

RaduBerinde committed:
opt: fix With and WithScan properties

The `With` and `WithScan` property builders start by copying the input
props and then correcting things. This is error-prone: we have in time
identified multiple things that were being incorrectly inherited.

This change rewrites the code to start with empty properties and then
fill things in explicitly (like most other operators). Known problems
that are fixed by this change:
 - `CanMutate` was incorrectly set for `WithScan`
 - Rule props like `PruneCols` and `WithUses` were incorrectly
   inherited by `With`.

Release note: None

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/40975)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @andreimatei, @jordanlewis, @justinj, and @rytaft)

___
*[pkg/sql/opt/memo/logical_props_builder.go, line 754 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/40975#-LpVp9TL4SJQUG_Te4aB:-LpYkQL096XZLDH7RxI5:b8zt95v) ([raw file](https://github.com/cockroachdb/cockroach/blob/8bd8d6c134d29fbabc8e945f0d439ebe9786a77b/pkg/sql/opt/memo/logical_props_builder.go#L754)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

Yeah, agreed, this has proven to be the wrong approach.
</blockquote></details>

Done. PTAL at the first commit.


<!-- Sent from Reviewable.io -->

justinj reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/40975)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @andreimatei, @jordanlewis, @justinj, and @rytaft)

___
*[pkg/sql/opt/memo/logical_props_builder.go, line 754 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/40975#-LpVp9TL4SJQUG_Te4aB:-LpYklYY5UbgrdDNWpuN:brfqlux) ([raw file](https://github.com/cockroachdb/cockroach/blob/8bd8d6c134d29fbabc8e945f0d439ebe9786a77b/pkg/sql/opt/memo/logical_props_builder.go#L754)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Done. PTAL at the first commit.
</blockquote></details>

<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> thanks!


<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 3 of 6 files at r3, 3 of 3 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/40975#-:-LpZdgSY4IWGcZFx5AxG:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @andreimatei, @jordanlewis, and @justinj)



<!-- Sent from Reviewable.io -->

RaduBerinde commented:
bors r+

andreimatei commented:
Once #41307 goes in, I think we can revert this one. Unions will be serialized and so they can have mutations under them now.

RaduBerinde commented:
Opened #41356 to revert the relevant commit.
